lib/improver/tests/spotdata_new/spotdata_new/test_NeighbourSelection.py

```diff
@@ -0,0 +1,726 @@
+        this case the z component should be a cos(45) component of the
+        semi-major axis of the globe defined in the global coordinate system.
+        The x and y coordinates should be cos(45) components of the remaining
+        cos(45) component of the semi-major axis."""
+
+        plugin = NeighbourSelection()
+        x_points = np.array([45])
+        y_points = np.array([45])
+        result = plugin.geocentric_cartesian(self.global_orography,
+                                             x_points, y_points)
+        radius = self.global_orography.coord_system().semi_major_axis
+        component = radius/np.sqrt(2.)
+        sub_component = component/np.sqrt(2.)
+        expected = [[sub_component, sub_component, component]]
+
+        self.assertArrayAlmostEqual(result, expected)
+
+    def test_negative_45_degrees_latitude_and_longitude(self):
+        """Test a (-45, -45) coordinate conversion to geocentric cartesian.
+        In this case the x is expected to remain positive, whilst y and z
+        become negative."""
+
+        plugin = NeighbourSelection()
+        x_points = np.array([-45])
+        y_points = np.array([-45])
+        result = plugin.geocentric_cartesian(self.global_orography,
+                                             x_points, y_points)
+        radius = self.global_orography.coord_system().semi_major_axis
+        component = radius/np.sqrt(2.)
+        sub_component = component/np.sqrt(2.)
+        expected = [[sub_component, -sub_component, -component]]
+
+        self.assertArrayAlmostEqual(result, expected)
+
+
+class Test_build_KDTree(Test_NeighbourSelection):
+
+    """Test construction of a KDTree with scipy."""
+
+    def test_basic(self):
+        """Test that the expected number of nodes are created and that a tree
+        is returned; this should be the lengths of the x and y coordinates
+        multiplied in the simple case."""
+
+        plugin = NeighbourSelection()
+        result, result_nodes = plugin.build_KDTree(self.region_land_mask)
+        expected_length = (self.region_land_mask.shape[0] *
+                           self.region_land_mask.shape[1])
+
+        self.assertEqual(result_nodes.shape[0], expected_length)
+        self.assertIsInstance(result, scipy.spatial.ckdtree.cKDTree)
+
+    def test_only_land(self):
+        """Test that the expected number of nodes are created and that a tree
+        is returned. In this case the number of nodes should be this should be
+        equal to the number of land points."""
+
+        plugin = NeighbourSelection(land_constraint=True)
+        result, result_nodes = plugin.build_KDTree(self.region_land_mask)
+        expected_length = np.nonzero(self.region_land_mask.data)[0].shape[0]
+
+        self.assertEqual(result_nodes.shape[0], expected_length)
+        self.assertIsInstance(result, scipy.spatial.ckdtree.cKDTree)
+
+
+class Test_select_minimum_dz(Test_NeighbourSelection):
```

Comment:
It might be worth seeing what happens in a tied situation, like where the islands either side have the same altitude, I think it will work, and this situation is unlikely to happen in reality.

Reply:
Done.